sdk/messaging/azservicebus/receiver.go

```diff
@@ -196,120 +200,118 @@ func (r *Receiver) ReceiveMessages(ctx context.Context, numMessages int, options
 		}
 	}
 
-	receiver, err := r.updateReceiver(ctx)
+	_, receiver, _, _, err := r.amqpLinks.Get(ctx)
 
 	if err != nil {
 		return nil, err
 	}
 
-	if err := receiver.IssueCredit(uint32(numMessages)); err != nil {
+	if err := receiver.IssueCredit(uint32(maxMessages)); err != nil {
 		return nil, err
 	}
 
 	var messages []*ReceivedMessage
-	var allMessagesReceivedError = errors.New("all messages received")
 
-	receiveCh, afterFirstMessageFn := startReceiverTimer(ropts.maxWaitTime, ropts.maxWaitTimeAfterFirstMessage)
-
-	listenCtx, cancelListenCtx := context.WithCancel(ctx)
-	defer cancelListenCtx()
+	ctx, cancel := context.WithTimeout(ctx, ropts.maxWaitTime)
+	defer cancel()
+
+	// the last "interesting" error to occur in the sequence. Used only if we end up receiving zero
+	// messages (otherwise, the user won't even know to look at the return value).
+	var fatalError error
+
+	// receive until we get our first interruption:
+	// - context is cancelled because the user cancelled it
+	// - context is cancelled because one of our timeouts hit
+	// - we get all the messages we asked for
+	for {
+		// TODO: should we do retries here? Or should we just surface the error close to immediately?
+		amqpMessage, err := receiver.Receive(ctx)
+
+		if err != nil {
+			// at the moment all errors from Receive() are fatal (if they're not context cancellation errors)
+			// for simplicity, we'll just store whatever we got and return it later if nothing else goes bad.
+			fatalError = err
+
+			if err != context.Canceled && err != context.DeadlineExceeded {
+				// link is dead. Close our local state so it can be recreated again on the next
+				// receiveMessage and return whatever we have (or this error if that's the best we can do)
+				if err := r.amqpLinks.Close(ctx, false); err != nil {
+					tab.For(ctx).Debug(fmt.Sprintf("failed when closing local links (not fatal): %s", err.Error()))
+				}
+
+				if len(messages) == 0 {
+					return nil, fatalError
+				} else {
+					// if they got _some_ messages we don't consider that fatal.
+					return messages, nil
+				}
+			}
+			break
+		}
 
-	listenHandle := r.receiver.Listen(listenCtx, internal.HandlerFunc(func(c context.Context, legacyMessage *internal.Message) error {
-		// NOTE: the invocation of this function from the AMQP library is single-threaded so
-		// no concurrency sync is required.
-		messages = append(messages, convertToReceivedMessage(legacyMessage))
+		messages = append(messages, newReceivedMessage(ctx, amqpMessage))
 
-		if len(messages) == numMessages {
-			return allMessagesReceivedError
+		if len(messages) == maxMessages {
+			break
 		}
 
 		if len(messages) == 1 {
-			afterFirstMessageFn()
-		}
-
-		return nil
-	}))
-
-	select {
-	case <-listenHandle.Done():
-		if listenHandle.Err() != allMessagesReceivedError {
-			err = listenHandle.Err()
+			go func() {
```

Comment:
What is this doing?  Comment would be good.

Reply:
Comment added